packages/builder/src/builderStore/store/index.js

```diff
@@ -454,3 +461,50 @@ const setMetadataProp = store => (name, prop) => {
     return s
   })
 }
+
+const storeComponentForCopy = store => (component, cut = false) => {
+  store.update(s => {
+    const copiedComponent = cloneDeep(component)
+    s.componentToPaste = copiedComponent
+    s.componentToPaste.isCut = cut
+    if (cut) {
+      const parent = getParent(s.currentPreviewItem.props, component._id)
+      parent._children = parent._children.filter(c => c._id !== component._id)
+      selectComponent(s, parent)
+    }
+
+    return s
+  })
+}
+
+const pasteComponent = store => (targetComponent, mode) => {
```

Comment:
This whole concept seems like a bit of a code smell - cloning the component and filtering the parents children, generating new IDs etc at the store level. There has to be a simpler solution.

It feels like there should just be one piece of state that controls this `cutComponent/copiedComponent` that has an ID/pointer to the currently cut/copied component, then other parts of the code might be updated accordingly.

Reply:
this has been in the codebase since we implemented copy/paste... i have only really moved it into the store, so i could reuse the function. Reluctant to go changing it now.

Reply:
In the scenarios
- Cut component
- Paste component
- Paste component again

OR 
- copy component
- paste component

... We need to generate new IDs, or else there will be a clash


In the scenarios

- cut component
- paste component

OR (less likely)
- copy component
- delete component
- paste component

Storing a reference to the original component does not work, as the original compnent no longer exists.

The existing solution covers all these scenarios just fine.

Reply:
Ah - it's a wider problem then. https://github.com/Budibase/budibase/issues/338 or some kind of abstraction around dealing with the component tree would solve a lot of these problems. If we don't use the DB, I think that a more OOP based approach might actually be a much better fit for this sort of stuff, so IDs can be generated in the `constructor` etc of component objects.